### PR TITLE
style: support mypy checking and update docstring style

### PR DESCRIPTION
**Pull request by Gaiejj**
## Description

- Modify omnisafe's code to a version that conforms to the mypy specification
- Replace the return value of ``_loss_pi`` from ``Tuple[torch.Tensor, dict]`` to ``torch.Tensor``. Use ``logger`` to record values inside ``_loss_pi`` instead of passing through dictionary ``info``.
- Activate ``make mypy`` in github ``.workflow``.
- Activate ``make pydocstyle`` in github ``.workflow``.
- Refactor the API docstring.
- Refactor the list-table to rst table.
- Make API docstring cover all public method and class.

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/OmniSafeAI/omnisafe/blob/HEAD/CONTRIBUTING.md) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [x] I have updated the documentation accordingly.
- [ ] I have reformatted the code using `make format`. (**required**)
- [x] I have checked the code using `make lint`. (**required**)
- [x] I have ensured `make test` pass. (**required**)

### CODE REVIEW

File: examples/benchmarks/run_experiment_grid.py [hunk 1]
```diff
@@ -22,23 +22,20 @@
...
 
 
 def train(
     exp_id: str,
     algo: str,
     env_id: str,
-    custom_cfgs: NamedTuple,
-) -> Tuple[float, float, float]:
+    custom_cfgs: dict,
+) -> tuple:
```
Comment (XuehaiPan) [inline examples/benchmarks/run_experiment_grid.py:32]: Should fill in the content type for generic containers. `dict` represents `dict[Any, Any]`.

File: omnisafe/common/simmer_agent.py [hunk 2]
```diff
@@ -41,14 +41,14 @@ def __init__(
...
-        self._budget_bound = budget_bound
+        self._budget_bound = torch.as_tensor(budget_bound, dtype=torch.float32)
         self._cfgs = cfgs
         # history
-        self._error_history = deque([], maxlen=self._history_len)
-        self._reward_history = deque([], maxlen=self._history_len)
-        self._state_history = deque([], maxlen=self._history_len)
-        self._action_history = deque([], maxlen=self._history_len)
-        self._observation_history = deque([], maxlen=self._history_len)
+        self._error_history: deque = deque([], maxlen=self._history_len)
```
Comment (XuehaiPan) [inline omnisafe/common/simmer_agent.py:47]:
```suggestion
        self._error_history: deque[torch.Tensor] = deque([], maxlen=self._history_len)
```

File: omnisafe/common/statistics_tools.py [hunk 3]
```diff
@@ -315,7 +326,7 @@ def variants(self, keys, vals):
         def check_duplicate(var):
             """Build the full nested dict version of var, based on key names."""
             new_var: dict = {}
-            unflatten_set = set()
+            unflatten_set: set = set()
```
Comment (XuehaiPan) [inline omnisafe/common/statistics_tools.py:329]: This variable seems unused.

File: omnisafe/utils/command_app.py [hunk 4]
```diff
@@ -99,11 +99,11 @@ def train(  # pylint: disable=too-many-arguments
     }
     keys = custom_cfgs[0::2]
     values = list(custom_cfgs[1::2])
-    custom_cfgs = dict(zip(keys, values))
-    custom_cfgs.update({'logger_cfgs:log_dir': os.path.join(log_dir, 'train')})
+    custom_cfgs_dict = dict(zip(keys, values))
```
Comment (XuehaiPan) [inline omnisafe/utils/command_app.py:102]: Useless `list` conversion. Also, Do we need to check that `len(custom_cfgs) % 2 == 0`?

File: omnisafe/adapter/onpolicy_adapter.py [hunk 5]
```diff
@@ -144,7 +143,6 @@ def _log_value(
         reward: torch.Tensor,
         cost: torch.Tensor,
         info: dict,
```
Comment (rockmagma02) [inline omnisafe/adapter/onpolicy_adapter.py:145]: maybe kwargs is useless?

File: omnisafe/utils/plotter.py [hunk 6]
```diff
@@ -61,14 +62,14 @@ def __init__(self) -> None:
...
-        sub_figures,
-        data,
-        xaxis='Steps',
-        value='Rewards',
-        condition='Condition1',
-        smooth=1,
-        **kwargs,
-    ):
+        sub_figures: np.ndarray,
+        data: list,
```
Comment (rockmagma02) [inline omnisafe/utils/plotter.py:66]: list[what]?

File: omnisafe/utils/plotter.py [hunk 7]
```diff
@@ -188,7 +189,13 @@ def get_datasets(self, logdir, condition=None):
...
 
-    def get_all_datasets(self, all_logdirs, legend=None, select=None, exclude=None):
+    def get_all_datasets(
+        self,
+        all_logdirs: list[str],
+        legend: list | None = None,
+        select: list | None = None,
+        exclude: list | None = None,
+    ) -> list[dict[str, Any]]:
         """
```
Comment (rockmagma02) [inline omnisafe/utils/plotter.py:199]: list[what]?

**Review by rockmagma02 — CHANGES_REQUESTED** [review]

File: omnisafe/adapter/onpolicy_adapter.py (hunk 5 above)
Comment (Gaiejj) [reply omnisafe/adapter/onpolicy_adapter.py:145]: I think so and I have deleted it.

File: omnisafe/utils/plotter.py (hunk 6 above)
Comment (Gaiejj) [reply omnisafe/utils/plotter.py:66]: Thanks for your hints. I will fill all those blank dict/tuple/list in future commits.

File: omnisafe/algorithms/algo_wrapper.py [hunk 8]
```diff
@@ -131,20 +131,20 @@ def _init_algo(self):
             cfgs=self.cfgs,
         )
 
-    def learn(self):
+    def learn(self) -> tuple:
```
Comment (Gaiejj) [inline omnisafe/algorithms/algo_wrapper.py:134]: specified the content in the tuple

File: omnisafe/algorithms/on_policy/second_order/cpo.py [hunk 9]
```diff
@@ -214,7 +215,7 @@ def _determine_case(
         q: torch.Tensor,
         r: torch.Tensor,
         s: torch.Tensor,
-    ) -> tuple(int, torch.Tensor, torch.Tensor):
+    ) -> tuple:
```
Comment (Gaiejj) [inline omnisafe/algorithms/on_policy/second_order/cpo.py:218]: specified the content in the tuple.

File: omnisafe/algorithms/on_policy/second_order/cpo.py [hunk 10]
```diff
@@ -270,7 +271,7 @@ def _step_direction(
         r: torch.Tensor,
         s: torch.Tensor,
         ep_costs: torch.Tensor,
-    ) -> tuple(torch.Tensor, ...):
+    ) -> tuple:
```
Comment (Gaiejj) [inline omnisafe/algorithms/on_policy/second_order/cpo.py:274]: The same.

File: omnisafe/algorithms/on_policy/second_order/cpo.py (hunk 9 above)
Comment (XuehaiPan) [reply omnisafe/algorithms/on_policy/second_order/cpo.py:218]:
```suggestion
    ) -> tuple[int, torch.Tensor, torch.Tensor]:
```

File: examples/benchmarks/run_experiment_grid.py [hunk 11]
```diff
@@ -14,31 +14,31 @@
...
 
 
 def train(
     exp_id: str,
     algo: str,
     env_id: str,
-    custom_cfgs: NamedTuple,
-) -> Tuple[float, float, float]:
+    custom_cfgs: dict[str, Any],
+) -> tuple[float, ...]:
```
Comment (XuehaiPan) [inline examples/benchmarks/run_experiment_grid.py:35]:
```suggestion
) -> tuple[float, float, float]:
```

File: omnisafe/adapter/offpolicy_adapter.py [hunk 12]
```diff
@@ -172,7 +171,6 @@ def _log_value(
         reward: torch.Tensor,
         cost: torch.Tensor,
         info: dict,
```
Comment (XuehaiPan) [inline omnisafe/adapter/offpolicy_adapter.py:173]: Need type hints for keys and values.

**Review by XuehaiPan — CHANGES_REQUESTED** [review]
You should revisit and update the whole project rather than trial-and-error with `mypy`.

Some common mistakes:

1. Missing type annotations for keys and values in `dict`.
2. Missing type annotations for items in `list` / `tuple`.
3. Misuse of `tuple[T, S, U]` and `tuple[T, ...]`.
4. Type annotation for attributes.

    ```python
    class SomeClass:
        _some_attr: SomeType  # annotate without an assignment

        ...
    ```

File: omnisafe/algorithms/algo_wrapper.py [hunk 13]
```diff
@@ -108,7 +115,7 @@ def _init_checks(self):
             self.env_id in support_envs()
         ), f"{self.env_id} doesn't exist. Please choose from {support_envs()}."
 
-    def _init_algo(self):
+    def _init_algo(self) -> None:
         """Init algo."""
         # Use number of physical cores as default.
```
Comment (zmsn-2077) [inline omnisafe/algorithms/algo_wrapper.py:120]: comment styple

File: omnisafe/adapter/early_terminated_adapter.py [hunk 14]
```diff
@@ -12,37 +12,48 @@
...
 
     def __init__(self, env_id: str, num_envs: int, seed: int, cfgs: Config) -> None:
         assert num_envs == 1, 'EarlyTerminatedAdapter only supports num_envs=1.'
 
         super().__init__(env_id, num_envs, seed, cfgs)
 
+        self._cost_limit: float
+        self._cost_logger: torch.Tensor
         self._cost_limit = cfgs.algo_cfgs.cost_limit
         self._cost_logger = torch.zeros(self._env.num_envs)
```
Comment (XuehaiPan) [inline omnisafe/adapter/early_terminated_adapter.py:38]: You do not need to split the type annotation and assignment into two lines.

```suggestion
        self._cost_limit: float = cfgs.algo_cfgs.cost_limit
        self._cost_logger: torch.Tensor = torch.zeros(self._env.num_envs)
```

If you define a attribute outside of `__init__`, you should annotate it with:

```python
class MyClass:
    _logger: Logger

    def __init__(self, ...):
        ...

        self._init_logger()

    def _init_logger(self):
        self._logger = Logger(self.config)
```

Comment (Gaiejj) [reply omnisafe/adapter/early_terminated_adapter.py:38]: I see, I just misunderstand ``# annotate without an assignment``, I will fix all mistakes like that in follow commits.

File: omnisafe/adapter/offpolicy_adapter.py [hunk 15]
```diff
@@ -208,7 +209,7 @@ def _reset_log(self, idx: int | None = None) -> None:
         """Reset log.
 
         Args:
-            idx (int | None): The index of the environment.
```
Comment (XuehaiPan) [inline omnisafe/adapter/offpolicy_adapter.py:211]: Follow the PyTorch guideline for type annotations in docstrings ([Docstring type formatting](https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md#docstring-type-formatting)).

File: omnisafe/adapter/online_adapter.py [hunk 16]
```diff
@@ -169,38 +172,42 @@ def observation_space(self) -> OmnisafeSpace:
...
+            info (dict[str, Any]): Some information logged by the environment.
         """
         return self._env.reset()
 
     def save(self) -> dict[str, torch.nn.Module]:
         """Save the environment.
 
         Returns:
-            Dict[str, torch.nn.Module]: the saved environment.
+            dict[str, torch.nn.Module]: The saved environment.
```
Comment (XuehaiPan) [inline omnisafe/adapter/online_adapter.py:211]: Remove type annotation in `Returns`. Only need a description.

File: omnisafe/adapter/online_adapter.py [hunk 17]
```diff
@@ -169,38 +172,42 @@ def observation_space(self) -> OmnisafeSpace:
...
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        dict[str, Any],
+    ]:
         """Run one timestep of the environment's dynamics using the agent actions.
 
         Args:
-            action (torch.Tensor): action.
+            action (torch.Tensor): action from the agent or random.
```
Comment (XuehaiPan) [inline omnisafe/adapter/online_adapter.py:186]:
```suggestion
            action (torch.Tensor): The action from the agent or random.
```

Should be a sentence and capitalize the first letter.

File: omnisafe/algorithms/off_policy/ddpg.py [hunk 18]
```diff
@@ -129,53 +129,43 @@ def _init(self) -> None:
...
+        +-------------------------+-------------------------------------------------------------------------+
+        | Metrics/EpLen           | Average length of the epoch.                                            |
+        +-------------------------+-------------------------------------------------------------------------+
+        | Metrics/TestEpCost      | Average cost of the evaluate epoch.                                     |
+        +-------------------------+-------------------------------------------------------------------------+
+        | Metrics/TestEpRet       | Average return of the evaluate epoch.                                   |
+        +-------------------------+-------------------------------------------------------------------------+
+        | Metrics/TestEpLen       | Average length of the evaluate epoch.                                   |
+        +-------------------------+-------------------------------------------------------------------------+
+        | Value/reward_critic     | Average value in :meth:`roll_out()` (from critic network) of the epoch. |
```
Comment (XuehaiPan) [inline omnisafe/algorithms/off_policy/ddpg.py:149]:
```suggestion
        | Value/reward_critic     | Average value in :meth:`roll_out` (from critic network) of the epoch. |
```

File: omnisafe/algorithms/on_policy/base/natural_pg.py [hunk 19]
```diff
@@ -47,24 +47,22 @@ class NaturalPG(PolicyGradient):
...
+        +=====================+====================================================+
+        | Misc/AcceptanceStep | The acceptance step size.                          |
+        +---------------------+----------------------------------------------------+
+        | Misc/Alpha          | :math:`\frac{\delta_{KL}}{xHx}` in original paper. |
+        +---------------------+----------------------------------------------------+
+        | Misc/FinalStepNorm  | The final step norm.                               |
+        +---------------------+----------------------------------------------------+
+        | Misc/gradient_norm  | The gradient norm.                                 |
+        +---------------------+----------------------------------------------------+
+        | Misc/xHx            | :math:`xHx` in original paper.                     |
```
Comment (XuehaiPan) [inline omnisafe/algorithms/on_policy/base/natural_pg.py:61]:
```suggestion
        | Misc/xHx            | :math:`xHx` in the original paper.                     |
```

File: omnisafe/adapter/saute_adapter.py [hunk 20]
```diff
@@ -27,41 +29,74 @@
...
+        - If the safety state is greater than 0, the reward is the original reward.
+        - If the safety state is less than 0, the reward is the unsafe reward (always 0 or less than 0).
+
+    OmniSafe provides two implementations of Saute RL: :class:`PPOSaute` and :class:`TRPOSaute`.
+
+    References:
+        - Title: Saute RL: Almost Surely Safe Reinforcement Learning Using State Augmentation
+        - Authors: Aivar Sootla, Alexander I. Cowen-Rivers, Taher Jafferjee, Ziyan Wang,
+            David Mguni, Jun Wang, Haitham Bou-Ammar.
+        - URL: https://arxiv.org/abs/2202.06558
```
Comment (zmsn-2077) [inline omnisafe/adapter/saute_adapter.py:48]: Should it be the same as
```
- URL: `DDPG <https://arxiv.org/abs/1509.02971>`_

```

File: omnisafe/adapter/simmer_adapter.py [hunk 21]
```diff
@@ -16,64 +16,93 @@
...
+        - If the safety state is greater than 0, the reward is the original reward.
+        - If the safety state is less than 0, the reward is the unsafe reward (always 0 or less than 0).
+
+    OmniSafe provides two implementations of Simmer RL: :class:`PPOSimmer` and :class:`TRPOSimmer`.
+
+    References:
+        - Title: Effects of Safety State Augmentation on Safe Exploration.
+        - Authors: Aivar Sootla, Alexander I. Cowen-Rivers, Taher Jafferjee, Ziyan Wang,
+            David Mguni, Jun Wang, Haitham Bou-Ammar.
+        - URL: https://arxiv.org/abs/2202.06558
```
Comment (zmsn-2077) [inline omnisafe/adapter/simmer_adapter.py:48]: Should it be the same as
```
- URL: `DDPG <https://arxiv.org/abs/1509.02971>`_

```

Comment (zmsn-2077) [inline omnisafe/adapter/simmer_adapter.py:48]:
```suggestion
        - URL: `Saute RL <https://arxiv.org/abs/2202.06558>`_
```

File: omnisafe/adapter/saute_adapter.py (hunk 20 above)
Comment (zmsn-2077) [inline omnisafe/adapter/saute_adapter.py:48]:
```suggestion
        - URL: `Saute RL <https://arxiv.org/abs/2202.06558>`_
```

**Review by zmsn-2077 — APPROVED** [review]
LGTM.

**Review by XuehaiPan — APPROVED** [review]

**Review by rockmagma02 — APPROVED** [review]

**Pull request merged**